File: frontend/src/lib/api/roleplay.ts
import { api } from "./client";
import type {
  Conversation,
  ConversationMessage,
  ConversationStartPayload,
  ConversationStartResponse,
  MessageSubmitResponse,
  RoleplayScenario,
} from "./types";

export const roleplayApi = {
  scenarios: () => api.get<RoleplayScenario[]>("/roleplay-scenarios"),
  listConversations: (userId: string, courseId: string) =>
    api.get<Conversation[]>(
      `/conversations?${new URLSearchParams({ user_id: userId, course_id: courseId })}`,
    ),
  startConversation: (payload: ConversationStartPayload) =>
    api.post<ConversationStartResponse>("/conversations", payload),
  messages: (conversationId: string, userId: string) =>
    api.get<ConversationMessage[]>(
      `/conversations/${conversationId}/messages?${new URLSearchParams({ user_id: userId })}`,
    ),
  sendMessage: (conversationId: string, userId: string, text: string) =>
    api.post<MessageSubmitResponse>(`/conversations/${conversationId}/messages`, {
      user_id: userId,
      text,
    }),
};

File: backend/app/schemas/roleplay.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import MessageRole


class RoleplayScenarioRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    slug: str
    setup_prompt: str
    order_index: int
    created_at: datetime


class ConversationStart(BaseModel):
    """Request body for POST /conversations."""

    user_id: uuid.UUID
    course_id: uuid.UUID
    scenario_id: uuid.UUID


class MessageSubmit(BaseModel):
    """Request body for POST /conversations/{id}/messages."""

    user_id: uuid.UUID
    text: str


class Correction(BaseModel):
    original: str
    corrected: str
    explanation: str


class ConversationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    course_id: uuid.UUID
    scenario_id: uuid.UUID
    created_at: datetime


class ConversationMessageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    conversation_id: uuid.UUID
    role: MessageRole
    text: str
    corrections: list[Correction] | None
    created_at: datetime


class ConversationStartResponse(BaseModel):
    conversation: ConversationRead
    messages: list[ConversationMessageRead]


class MessageSubmitResponse(BaseModel):
    user_message: ConversationMessageRead
    assistant_message: ConversationMessageRead
